pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -986,6 +986,14 @@ func (p *PriorityQueue) Update(logger klog.Logger, oldPod, newPod *v1.Pod) error
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	if p.isSchedulingQueueHintEnabled {
+		// the inflight pod will be requeued using the latest version from the informer cache, which matches what the event delivers.
+		if _, ok := p.inFlightPods[newPod.UID]; ok {
+			logger.V(6).Info("The pod doesn't be queued for now because it's being scheduled and will be queued back if necessary", "pod", klog.KObj(newPod))
+			return nil
+		}
+	}
```

Comment:
The current implementation has possible problematic scenarios.
Example:
1. A Pod is created and put in the queue.
2. The Pod starts to be scheduled (= Pop()ed out from the queue)
3. The Pod is updated and got a new toleration, but currently the scheduling queue doesn't receive Pod/Update when unscheduled Pods are updated. 
5. All Nodes in the cluster are tainted at the moment (wow, what the hell!) and the Pod is rejected by TaintToleration plugin.
6. The Pod comes back to the scheduling queue and put in unschedQ.
7. Even though this Pod already gets a toleration for the taint, but this Pod will never be moved to activeQ/backoffQ, until [flush](https://github.com/kubernetes/kubernetes/blob/95e16917f78310c16c81437c523d66abf716aeba/pkg/scheduler/internal/queue/scheduling_queue.go#L860-L879).

Regarding:
> currently the scheduling queue doesn't receive Pod/Update when unscheduled Pods are updated. 

See [updatePodInSchedulingQueue](https://github.com/kubernetes/kubernetes/blob/6acb46b62005e02ba196e488550ab725429323e1/pkg/scheduler/eventhandlers.go#L134-L156); we don't call `MoveAllToActiveOrBackoffQueue` (which records events into `inFlightEvents`) when unscheduled Pods are updated because we (mistakenly) assume events to unscheduled Pods never make any Pods schedulable.
 
So, that's the core problem to solve.

The ideal/general fix is to call `MoveAllToActiveOrBackoffQueue` with Pod/Update for unsched Pods update too, but that means the queue would start to get more events suddenly, which we would like to be careful for the sake of https://github.com/kubernetes/kubernetes/issues/120622. 

Sorry I put a long comment here; so, as a conclusion, my proposal for this problem is to put a Pod/Update event into `inFlightEvents` here, as a temporal solution _for now_, until the ideal fix arrives.

```suggestion
	if p.isSchedulingQueueHintEnabled {
		// the inflight pod will be requeued using the latest version from the informer cache, which matches what the event delivers.
		if _, ok := p.inFlightPods[newPod.UID]; ok {
			logger.V(6).Info("The pod doesn't be queued for now because it's being scheduled and will be queued back if necessary", "pod", klog.KObj(newPod))
			
			// Record this update as Pod/Update because 
			// this update may make the Pod schedulable in case it gets rejected and comes back to the queue.
			// We can clean it up once we change updatePodInSchedulingQueue to call MoveAllToActiveOrBackoffQueue.
			// See <GitHub link to this comment of mine> for more context.
			p.inFlightEvents.PushBack(&clusterEvent{
				event:  UnscheduledPodUpdate,
				oldObj: oldPod,
				newObj: newPod,
			})
			
			return nil
		}
	}
```

Reply:
Thanks for your detailed explanation.
I want to confirm that if `determineSchedulingHintForInFlightPod` function could make the pod in your case above be requeued immediately instead of waiting until flush.
https://github.com/kubernetes/kubernetes/blob/5ec31e84d6c525c173906b1497ee6f075c1926e9/pkg/scheduler/internal/queue/scheduling_queue.go#L705

Reply:
Yes, that is how my proposed fix should work; In my example scenario above, Pod/update Qhint in tainttoleration plugin should return Queue in determineSchedulingHintForInFlightPod and the Pod is requeued to backoffQ.

Please make sure if it really works by adding a unit test!

Reply:
> Please make sure if it really works by adding a unit test!

It's really hard to simulate the scenario in a unit/intergration test that a Pod is being scheduled while all nodes are tainted, and at the same time, the Pod receives an update event.

Reply:
Unit tests should be easily enough. (If you're doing integ tests, yes it's hard.)

1. `Add()` a pod to the queue (simulate pod creation)
2. `Pop()` the pod from the queue (simulate the start of scheduling)
3. `Update()` the Pod (simulate the pod getting updated while scheduling)
4. `AddUnschedulableIfNotPresent()` the Pod (simulate the Pod coming back to the scheduling queue)